File: pythonCode/Huber/getPerformanceProfiles.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Mar 29 12:18:58 2022

@author: cassiebuhler
"""
import numpy as np
from collections import Counter
import matplotlib.pyplot as plt


def getRatio(mat, isIter):
    best = mat.min(axis=1)
    ratios = mat/best[:,None]
    maxIter = 1000
    if isIter:
        ratios[mat == maxIter] = np.nan; #if it doesn't solve, set ratio to nan 
    return ratios

def performanceProf(ratios):
  ratio = sorted(ratios)
  counts = Counter(ratio)
  counts = sorted(counts.items(), key=lambda x: x[0])
  probs = []
  taus = []
  prob = 0
  for count in counts:
    prob += count[1]/len(ratio)
    probs.append(prob)
    taus.append(count[0])
  return np.vstack((taus,probs))


def getPerformanceProfiles(iters,time):
    ratios_time = getRatio(time,0) 
    ratios_iters = getRatio(iters,1)

    ppIters_cgNoCubic = performanceProf(ratios_iters[:,0])
    ppIters_cgCubic = performanceProf(ratios_iters[:,1])

    ppTime_cgNoCubic = performanceProf(ratios_time[:,0])
    ppTime_cgCubic = performanceProf(ratios_time[:,1])
    plt.figure(0)
    plt.title("Performance Profile: Iterations") 
    plt.xlabel(r"Tau")
    plt.ylabel("Probability")
    plt.plot(ppIters_cgNoCubic[0,:],ppIters_cgNoCubic[1,:], label = 'No Cubic', marker='o', linestyle='dashed', markersize = 8,linewidth = 2)
    plt.plot(ppIters_cgCubic[0,:],ppIters_cgCubic[1,:], label = 'With Cubic', marker='o', linestyle='dashed', markersize = 8,linewidth = 2)
    plt.legend()
    plt.ylim([0,1.1])

    
    plt.figure(1)
    plt.title("Performance Profile: Time") 
    plt.xlabel(r"Tau")
    plt.ylabel("Probability")
    plt.plot(ppTime_cgNoCubic[0,:],ppTime_cgNoCubic[1,:], label = 'No Cubic', marker='o', linestyle='dashed', markersize = 8,linewidth = 2)
    plt.plot(ppTime_cgCubic[0,:],ppTime_cgCubic[1,:], label = 'With Cubic', marker='o', linestyle='dashed', markersize = 8,linewidth = 2)
    plt.legend()
    plt.ylim([0,1.1])
    return

File: matlabCode/groupLASSO/getPerformanceProfiles.m
function getPerformanceProfiles(time,iters,status)

ratios_time = getRatio(time,status); %compute ratio for performance profile
ratios_iters = getRatio(iters(:,1:2),status(:,1:2));%exclude admm from iterations pp

ppIters_cgNoCubic = performanceProf(ratios_iters(:,1));
ppIters_cgCubic = performanceProf(ratios_iters(:,2));

ppTime_cgNoCubic = performanceProf(ratios_time(:,1));
ppTime_cgCubic = performanceProf(ratios_time(:,2));
ppTime_admm = performanceProf(ratios_time(:,3));

figure
title("Performance Profile: Time",'fontsize',18)
xlabel("Tau",'fontsize',16)
ylabel("Probability",'fontsize',16)
hold on
plot(ppTime_cgNoCubic(:,1),ppTime_cgNoCubic(:,2),'*-','markersize',8,'LineWidth',2)
plot(ppTime_cgCubic(:,1),ppTime_cgCubic(:,2),'*-','markersize',8,'LineWidth',2)
plot(ppTime_admm(:,1),ppTime_admm(:,2),'*-','markersize',8,'LineWidth',2)
legend({'CG without Cubic Reg','CG with Cubic Reg','ADMM'},'location','southeast','fontsize',14)
ylim([0,1])


figure
title("Performance Profile: Iterations",'fontsize',18)
xlabel("Tau",'fontsize',16)
ylabel("Probability",'fontsize',16)
hold on
plot(ppIters_cgNoCubic(:,1),ppIters_cgNoCubic(:,2),'*-','markersize',8,'LineWidth',2)
plot(ppIters_cgCubic(:,1),ppIters_cgCubic(:,2),'*-','markersize',8,'LineWidth',2)
legend({'CG without Cubic Reg','CG with Cubic Reg'},'location','southeast','fontsize',14)
ylim([0,1.05])

end

function ratios = getRatio(vec, status)
best = min(vec,[],2);
ratios = vec./best; %compute ratios
ratios(status ~= 0) = NaN; %if it doesn't solve, set ratio to nan
end

function performanceData = performanceProf(ratios) %compute performance prof
ratio = sort(ratios);
uniqueVals = unique(ratio);
counts=[];
for i=1:length(uniqueVals)
    counts(i) = sum(ratio == uniqueVals(i));
end
probs = [];
taus = [];
prob = 0;
for j=1:length(counts)
    prob = prob + counts(j)/length(ratio);
    probs(j) = prob;
    taus(j) = uniqueVals(j);
end
performanceData = [taus;probs]';
end

